controllers/packetmachine_controller.go

```diff
@@ -208,15 +210,22 @@ func (r *PacketMachineReconciler) reconcile(ctx context.Context, machineScope *s
 			return ctrl.Result{}, err
 		}
 	}
+
+	if dev == nil {
```

Comment:
This could use a comment block.  Help me understand this. We did not find a device with the `providerID` (which was presumably the UUID we received on a previous build attempt, success or failure, or externally deleted), and now in this block we are going to try to adopt any existing device that carries the tags we would have placed on an attempt to create this device.

Is this because the create call can error before returning the UUID, yet it may succeed, and we want to try to salvage the creation?

Reply:
I think the logic for this is off. I don't believe `create` can be reached:

```
if id {
 get device or die
}

if (no id, inferred) {
  get device by tags or die
}

if (no id, same as last comparisson) {
  // unreachable
  create device
}

if device /* get by id or tag succeeded */ {
  ...
}
```

Reply:
I think we could simplify this:
 
```
if id {
  get device or die
} else {
  get device by tag or die
}

if ! device {
  create device,  or die
}

device is guaranteed
sync
```

Reply:
Updated comments here, I believe the logic isn't currently too far removed from the pseudo code you suggested, just done without the use of `else`

> I think we could simplify this:
> 
> ```
> if id {
>   get device or die
> } else {
>   get device by tag or die
> }
> 
> if ! device {
>   create device or die
> }
> 
> // device is guaranteed
> sync
> ```

Reply:
I was way wrong here :-)

I was thinking an error would be returned by `GetDeviceByTags`. I didn't realize `GetDeviceByTags` was a `p.Devices.List` call that wouldn't surface a 404 or any other errors in a "not found" situation.  

This looks good.